data/abi/abi_encode.go

```diff
@@ -0,0 +1,444 @@
+	case Tuple:
+		return tupleEncoding(v)
+	default:
+		return []byte{}, fmt.Errorf("bruh you should not be here in encoding: unknown type error")
+	}
+}
+
+// compressMultipleBool compress consecutive bool values into a byte in ABI tuple/array value.
+func compressMultipleBool(valueList []Value) (uint8, error) {
+	var res uint8 = 0
+	if len(valueList) > 8 {
+		return 0, fmt.Errorf("value list passed in should be less than length 8")
+	}
+	for i := 0; i < len(valueList); i++ {
+		if valueList[i].ABIType.enumIndex != Bool {
+			return 0, fmt.Errorf("bool type not matching in compressMultipleBool")
+		}
+		boolVal, err := valueList[i].GetBool()
+		if err != nil {
+			return 0, err
+		}
+		if boolVal {
+			res |= 1 << uint(7-i)
+		}
+	}
+	return res, nil
+}
+
+// tupleEncoding encodes an ABI value of tuple type into an ABI encoded byte string.
+func tupleEncoding(v Value) ([]byte, error) {
+	if v.ABIType.enumIndex != Tuple {
+		return []byte{}, fmt.Errorf("type not supported in tupleEncoding")
+	}
+	if len(v.ABIType.childTypes) >= (1 << 16) {
+		return []byte{}, fmt.Errorf("value abi type exceed 2^16")
+	}
+	tupleElems := v.value.([]Value)
+	if len(tupleElems) != len(v.ABIType.childTypes) {
+		return []byte{}, fmt.Errorf("tuple abi child type number unmatch with tuple argument number")
+	}
+
+	heads := make([][]byte, len(v.ABIType.childTypes))
+	tails := make([][]byte, len(v.ABIType.childTypes))
+	isDynamicIndex := make(map[int]bool)
+
+	for i := 0; i < len(v.ABIType.childTypes); i++ {
+		if tupleElems[i].ABIType.IsDynamic() {
+			headsPlaceholder := []byte{0x00, 0x00}
+			heads[i] = headsPlaceholder
+			isDynamicIndex[i] = true
+			tailEncoding, err := tupleElems[i].Encode()
+			if err != nil {
+				return []byte{}, err
+			}
+			tails[i] = tailEncoding
+		} else {
+			if tupleElems[i].ABIType.enumIndex == Bool {
+				// search previous bool
+				before := findBoolLR(v.ABIType.childTypes, i, -1)
+				// search after bool
+				after := findBoolLR(v.ABIType.childTypes, i, 1)
+				// append to heads and tails
+				if before%8 != 0 {
+					return []byte{}, fmt.Errorf("expected before has number of bool mod 8 = 0")
+				}
+				if after > 7 {
+					after = 7
+				}
+				compressed, err := compressMultipleBool(tupleElems[i : i+after+1])
+				if err != nil {
+					return []byte{}, err
+				}
+				heads[i] = []byte{compressed}
+				i += after
+			} else {
+				encodeTi, err := tupleElems[i].Encode()
+				if err != nil {
+					return []byte{}, err
+				}
+				heads[i] = encodeTi
+			}
+			isDynamicIndex[i] = false
+		}
+	}
+
+	// adjust heads for dynamic type
+	headLength := 0
+	for _, headTi := range heads {
+		headLength += len(headTi)
+	}
+
+	tailCurrLength := 0
+	for i := 0; i < len(heads); i++ {
+		if isDynamicIndex[i] {
+			headValue := headLength + tailCurrLength
+			if headValue >= (1 << 16) {
+				return []byte{}, fmt.Errorf("encoding error: byte length exceed 2^16")
+			}
+			binary.BigEndian.PutUint16(heads[i], uint16(headValue))
```

Comment:
could you please clarify a bit this heads[i] encoding, can't match to the spec `head(x[i]) = enc(len( head(x[1]) ... head(x[N]) tail(x[1]) ... tail(x[i-1]) ))`